<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Ralph   University of Oregon   Institute of Ecology and Evolution">
  <title>New tools for popgen simulation and analysis:   What’s possible?</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="https://unpkg.com/reveal.js@^4//dist/reset.css">
  <link rel="stylesheet" href="https://unpkg.com/reveal.js@^4//dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <link rel="stylesheet" href="https://unpkg.com/reveal.js@^4//dist/theme/simple.css" id="theme">
  <style type="text/css">

  .reveal { font-size: 30px; }

  .reveal h1 { font-size: 1.5em; } 

  .reveal h2 { font-size: 1.2em; } 

  .reveal .slides { text-align: left; }

  .reveal .slides figure { text-align: center; }

  .centered p { text-align: center; }

  /* apply this to the columns div */
  div.vcentered { display: flex; align-items: center; }

  .caption p { font-size: 75%; }

  .greyed { color: #4444; }

  div.floatright { float: right; text-align: right; }

  .reveal figcaption { display: none; }

  .reveal img { border: none; }

  .reveal section img { border: none; }

  .MathJax_Display { margin: 0; }

  img.finger { vertical-align: text-top; }

  .smaller { font-size: 50%; }
  .smallish { font-size: 80%; }

  .hide { opacity: 0; }

  </style>

  <div style="display: none">
  \[
  \newcommand{\E}{\mathbb{E}}
  \renewcommand{\P}{\mathbb{P}}
  \DeclareMathOperator{\var}{var}
  \]
  </div>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">New tools for popgen simulation and analysis: <br/> What’s possible?</h1>
  <p class="author">Peter Ralph <br/> University of Oregon <br/> Institute of Ecology and Evolution</p>
  <p class="date">Evolution 2021</p>
</section>

<section>
<section id="overview-of-simulators" class="title-slide slide level1">
<h1>Overview of simulators</h1>
<!--
- overview of simulators and development
- background on forwards vs coalescent
- bullet points of features
- tree sequences: adding neutral mutations afterwards
- benchmarking for "plain vanilla": how many individuals and how long a genome?
- benchmarking for adding selection to (4) as a function of how much selection
- benchmarking for spatial nonWF: as a function of neighborhood size
- recapitation best practices
-->
<p><em>In this talk:</em></p>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><a href="https://tskit.dev/msprime">msprime</a>: a coalescent simulator</li>
<li><a href="https://messerlab.org/SLiM">SLiM</a>: a forwards simulator</li>
</ul>
<p>Other good ones:</p>
<ul>
<li><a href="https://tskit.dev/fwdpy11">fwdpy11</a></li>
<li><a href="http://www1.montpellier.inra.fr/CBGP/software/gspace/download.html">Gspace</a></li>
<li><a href="https://geonomics.readthedocs.io/en/latest/">geonomics</a></li>
<li><a href="https://github.com/RemiMattheyDoret/SimBit">simbit</a></li>
<li><a href="http://cmpg.unibe.ch/software/fastsimcoal2/">fastsimcoal</a></li>
<li><a href="http://sfscode.sourceforge.net/SFS_CODE/index/index.html">SFS_CODE</a></li>
</ul>
</div><div class="column" style="width:50%;">
<figure>
<img data-src="figs/msprime-logo.png" style="width:80.0%" alt="msprime logo" /><figcaption aria-hidden="true">msprime logo</figcaption>
</figure>
<figure>
<img data-src="figs/slim_logo.svg" style="width:60.0%" alt="SLiM logo" /><figcaption aria-hidden="true">SLiM logo</figcaption>
</figure>
</div>
</div>
</section>
<section id="development-philosophy" class="slide level2">
<h2>Development philosophy</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>open, welcoming, supportive</li>
<li>well-documented</li>
<li>reliable, reproducible</li>
<li>backwards compatible</li>
</ul>
</div><div class="column" style="width:50%;">
<figure>
<img data-src="figs/tskit_logo.png" style="width:80.0%" alt="tskit logo" /><figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<p><em>tskit</em>: the tree sequence toolkit</p>
<div class="floatright">
<p><a href="https://tskit.dev">https://tskit.dev</a></p>
</div>
</div>
</div>
</section>
<section id="forwards-or-backwards" class="slide level2">
<h2>Forwards or backwards?</h2>
<p><img data-src="figs/simulation_types.ink.svg" style="width:100.0%" /></p>
</section>
<section id="forwards-or-backwards-1" class="slide level2">
<h2>Forwards or backwards?</h2>
<p>Do your digital organisms:</p>
<ul>
<li class="fragment">have at most one site under selection?</li>
<li class="fragment">live in a collection of randomly-mating populations?</li>
<li class="fragment">not need some specific life cycle?</li>
</ul>
<div class="fragment">
<p>If so, then <em>coalescent simulation</em> is the way to go!</p>
</div>
</section></section>
<section>
<section id="msprime" class="title-slide slide level1">
<h1>msprime</h1>
<figure>
<img data-src="figs/msprime-logo.png" alt="msprime logo" /><figcaption aria-hidden="true">msprime logo</figcaption>
</figure>
<div class="floatright">
<figure>
<img data-src="figs/jerome.jpeg" style="width:50.0%" alt="jerome kelleher" /><figcaption aria-hidden="true">jerome kelleher</figcaption>
</figure>
<div class="caption">
<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, &amp; McVean</a></p>
</div>
</div>
</section>
<section id="msprime-v1.0" class="slide level2">
<h2>msprime v1.0</h2>
<figure>
<img data-src="figs/msp_folks.png" alt="msprime collaborators" /><figcaption aria-hidden="true">msprime collaborators</figcaption>
</figure>
</section>
<section id="new-features" class="slide level2">
<h2>New features:</h2>
<div class="columns">
<div class="column">
<ul>
<li><span class="math inline">\(k\)</span>-ploid individuals, finite sites</li>
</ul>
<div class="hide">
<ul>
<li>recombination rate maps</li>
<li>gene conversion</li>
<li>nicer demographic model specification</li>
<li>mutation rate maps</li>
<li>quick analysis</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/example1.png" /></p>
</div>
</div>
</section>
<section id="new-features-1" class="slide level2">
<h2>New features:</h2>
<div class="columns">
<div class="column">
<ul>
<li><span class="math inline">\(k\)</span>-ploid individuals, finite sites</li>
<li>recombination rate maps</li>
</ul>
<div class="hide">
<ul>
<li>gene conversion</li>
<li>nicer demographic model specification</li>
<li>mutation rate maps</li>
<li>quick analysis</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/example2.png" /></p>
</div>
</div>
</section>
<section id="new-features-2" class="slide level2">
<h2>New features:</h2>
<div class="columns">
<div class="column">
<ul>
<li><span class="math inline">\(k\)</span>-ploid individuals, finite sites</li>
<li>recombination rate maps</li>
<li>gene conversion</li>
</ul>
<div class="hide">
<ul>
<li>nicer demographic model specification</li>
<li>mutation rate maps</li>
<li>quick analysis</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/example3.png" /></p>
</div>
</div>
</section>
<section id="new-features-3" class="slide level2">
<h2>New features:</h2>
<div class="columns">
<div class="column">
<ul>
<li><span class="math inline">\(k\)</span>-ploid individuals, finite sites</li>
<li>recombination rate maps</li>
<li>gene conversion</li>
<li>nicer demographic model specification</li>
</ul>
<div class="hide">
<ul>
<li>mutation rate maps</li>
<li>quick analysis</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/example4.png" /></p>
</div>
</div>
</section>
<section id="new-features-4" class="slide level2">
<h2>New features:</h2>
<div class="columns">
<div class="column">
<ul>
<li><span class="math inline">\(k\)</span>-ploid individuals, finite sites</li>
<li>recombination rate maps</li>
<li>gene conversion</li>
<li>nicer demographic model specification</li>
<li>mutation rate maps</li>
</ul>
<div class="hide">
<ul>
<li>quick analysis</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/example5.png" /></p>
</div>
</div>
</section>
<section id="new-features-5" class="slide level2">
<h2>New features:</h2>
<div class="columns">
<div class="column">
<ul>
<li><span class="math inline">\(k\)</span>-ploid individuals, finite sites</li>
<li>recombination rate maps</li>
<li>gene conversion</li>
<li>nicer demographic model specification</li>
<li>mutation rate maps</li>
<li>quick analysis</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/example6.png" /></p>
</div>
</div>
</section>
<section id="ancestry-models" class="slide level2">
<h2>Ancestry models</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>“the” coalescent</li>
<li>discrete-time Wright-Fisher</li>
<li>selective sweeps</li>
<li>multiple mergers</li>
</ul>
</div><div class="column" style="width:60%;">
<div class="small">
<pre><code>sweep_model = msprime.SweepGenicSelection(
    position=2.5e4, s=0.01,
    start_frequency=0.5e-4, end_frequency=0.99, dt=1e-6)
sts = msprime.sim_ancestry(9,
    model=[sweep_model, msprime.StandardCoalescent()],
    population_size=1e4, recombination_rate=1e-8, sequence_length=5e4)</code></pre>
</div>
</div>
</div>
<p><img data-src="figs/sweep_trees.svg" /></p>
</section>
<section id="mutation-models" class="slide level2">
<h2>Mutation models</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>infinite sites/alleles</li>
<li>nucleotides</li>
<li>amino acids</li>
<li>arbitrary Markovian models</li>
</ul>
</div><div class="column" style="width:60%;">
<div class="small">
<pre><code>dem = msprime.Demography.from_species_tree(
   &quot;((A:900,B:900)ab:100,C:1000)abc;&quot;,
   initial_size=1e3)
samples = {&quot;A&quot;: 2, &quot;B&quot;: 1, &quot;C&quot;: 1}
ts = msprime.sim_ancestry(
   8, demography=dem, sequence_length=5e4,
   recombination_rate=1e-8
)
mts = msprime.sim_mutations(ts, rate=1e-7)
mts.draw_svg()</code></pre>
</div>
</div>
</div>
<p><img data-src="figs/mutated_trees.svg" /></p>
</section></section>
<section>
<section id="slim" class="title-slide slide level1" data-background-image="figs/slim_screenshot.png" data-background-position="left" data-background-size="100%">
<h1 data-background-image="figs/slim_screenshot.png" data-background-position="left" data-background-size="100%">SLiM</h1>

</section>
<section id="an-eco-evolutionary-simulator" class="slide level2">
<h2>An eco-evolutionary simulator</h2>
<div class="columns">
<div class="column">
<ul>
<li>everything msprime can</li>
<li>ecological dynamics with “non-Wright-Fisher” models</li>
<li>populations in continuous, heterogeneous geography</li>
<li>sex chromosomes, haplodiploidy</li>
<li>complex traits</li>
<li>context-dependent mutations</li>
</ul>
</div><div class="column" style="width:50%;">
<figure>
<img data-src="figs/slim_logo.svg" style="width:80.0%" alt="SLiM logo" /><figcaption aria-hidden="true">SLiM logo</figcaption>
</figure>
<figure>
<img data-src="figs/ben-haller.jpg" style="width:30.0%" alt="Ben Haller" /><figcaption aria-hidden="true">Ben Haller</figcaption>
</figure>
<div class="caption floatright">
<p>Ben Haller</p>
</div>
</div>
</div>
</section>
<section id="section" class="slide level2" data-background-image="figs/slim_manual.png" data-background-position="left" data-background-size="50%">
<h2 data-background-image="figs/slim_manual.png" data-background-position="left" data-background-size="50%"></h2>
<div class="columns">
<div class="column" style="width:60%;">

</div><div class="column" style="width:40%;">
<p><strong>Getting started:</strong></p>
<ol type="1">
<li>read the introduction of the SLiM manual</li>
<li>find a recipe that’s close to what you want</li>
<li>open up the GUI and try it</li>
<li>print stuff in the console</li>
<li>add in other bits</li>
<li>take a workshop!</li>
</ol>
</div>
</div>
</section></section>
<section>
<section id="tree-sequences" class="title-slide slide level1">
<h1>tree sequences</h1>
<figure>
<img data-src="figs/tskit_logo.png" style="width:30.0%" alt="tskit logo" /><figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<div class="floatright">
<figure>
<img data-src="figs/jerome.jpeg" style="width:50.0%" alt="jerome kelleher" /><figcaption aria-hidden="true">jerome kelleher</figcaption>
</figure>
<div class="caption">
<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, &amp; McVean</a></p>
</div>
</div>
</section>
<section class="slide level2">

<figure>
<img data-src="figs/tsk_folks.png" alt="tskit contributors" /><figcaption aria-hidden="true">tskit contributors</figcaption>
</figure>
</section>
<section id="the-tree-sequence" class="slide level2">
<h2>The tree sequence</h2>
<p><video data-src="figs/MCEB-final_animation.mp4" controls=""><a href="figs/MCEB-final_animation.mp4">Video</a></video></p>
<div class="caption">
<p><em>video credit: Yan Wong</em></p>
</div>
</section>
<section id="benefits" class="slide level2">
<h2>Benefits</h2>
<div class="columns">
<div class="column">
<ul>
<li>extremely efficient for large simulations</li>
<li>retains genotypes <em>and</em> genealogical history</li>
</ul>
<p>Interoperable: now supported by</p>
<ul>
<li><a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a></li>
<li><a href="https://messerlab.org/SLiM">SLiM</a></li>
<li><a href="https://molpopgen.github.io/fwdpy11/intro.html">fwdpy11</a></li>
<li><a href="http://www1.montpellier.inra.fr/CBGP/software/gspace/download.html">Gspace</a></li>
<li><a href="https://geonomics.readthedocs.io/en/latest/">geonomics</a></li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/fire_trees.svg" /></p>
</div>
</div>
</section>
<section id="post-hoc-mutations" class="slide level2">
<h2>Post-hoc mutations</h2>
<figure>
<img data-src="figs/add_mutations.svg" alt="diagram of adding mutations to a tree sequence" /><figcaption aria-hidden="true">diagram of adding mutations to a tree sequence</figcaption>
</figure>
</section>
<section id="recapitation" class="slide level2">
<h2>Recapitation</h2>
<figure>
<img data-src="figs/recapitation.svg" alt="diagram of recapitation" /><figcaption aria-hidden="true">diagram of recapitation</figcaption>
</figure>
</section></section>
<section>
<section id="runtime" class="title-slide slide level1">
<h1>Runtime</h1>

</section>
<section id="considerations" class="slide level2">
<h2>Considerations</h2>
<ul>
<li><span class="math inline">\(N\)</span> = population size</li>
<li><span class="math inline">\(L\)</span> = genome length</li>
<li>sample size (doesn’t matter much)</li>
<li>number of generations (SLiM only)</li>
<li>selection</li>
<li>geography</li>
<li>adding neutral mutations (nearly instant)</li>
</ul>
</section>
<section id="msprime-1000-samples" class="slide level2">
<h2>msprime: 1000 samples</h2>
<p><img data-src="benchmarking/msprime_times_small.png" style="width:80.0%" /></p>
<p><em>takeaway:</em> hundreds of thousands of megabases takes seconds</p>
</section>
<section id="msprime-1000-samples-1" class="slide level2">
<h2>msprime: 1000 samples</h2>
<p><img data-src="benchmarking/msprime_times.png" style="width:80.0%" /></p>
<p><em>takeaway:</em> hundreds of thousands of megabases takes seconds</p>
</section>
<section id="basic-demography-slim" class="slide level2">
<h2>basic demography: SLiM</h2>
<p><img data-src="benchmarking/run_wf_neutral.speeds.png" style="width:80.0%" /></p>
<p><em>takeaway:</em> linear in population size</p>
</section>
<section id="basic-demography-slim-1" class="slide level2">
<h2>Basic demography: SLiM</h2>
<p><img data-src="benchmarking/run_wf_neutral.speeds_perN.png" style="width:80.0%" /></p>
<p><em>takeaway:</em> seconds per thousand individuals per thousand generations</p>
</section>
<section id="selection-slim-total-rate-10-10" class="slide level2">
<h2>Selection: SLiM, total rate <span class="math inline">\(10^{-10}\)</span></h2>
<p><img data-src="benchmarking/run_wf_selection.speeds_perN.png" style="width:80.0%" /></p>
<p><em>takeaway:</em> similar, but slower by a factor of 3 for lots of positive mutations</p>
</section>
<section id="spatial-simulations-slim" class="slide level2">
<h2>Spatial simulations: SLiM</h2>
<p><img data-src="benchmarking/run_spatial.speeds.png" style="width:80.0%" /></p>
<p><em>takeaway:</em> 3x slower than genomes! Scales with neighborhood size (<span class="math inline">\(\sigma^2\)</span>).</p>
</section>
<section id="how-long-do-i-run-it-for" class="slide level2">
<h2>How long do I run it for?</h2>
<ol type="1">
<li class="fragment">Until equilibrium. (4N? 20N?)</li>
<li class="fragment">If that’s too long, for a “while”, and recapitate.</li>
<li class="fragment">Your results shouldn’t depend too much on how you do it.</li>
</ol>
<div class="fragment">
<p>Big picture: how accurate do you think your demographic model reflects 2N generations ago, really?</p>
</div>
</section></section>
<section>
<section id="thanks" class="title-slide slide level1">
<h1>Thanks!</h1>
<div class="columns">
<div class="column">
<figure>
<img data-src="figs/bdi.png" style="width:30.0%" alt="BDI" /><figcaption aria-hidden="true">BDI</figcaption>
</figure>
<figure>
<img data-src="figs/nih.jpeg" alt="NHGRI" /><figcaption aria-hidden="true">NHGRI</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<figure>
<img data-src="figs/tskit_logo.png" style="width:50.0%" alt="tskit logo" /><figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<ul>
<li>Jerome Kelleher</li>
<li>Ben Haller</li>
<li>Ben Jeffery</li>
<li>Yan Wong</li>
<li>Murillo Rodrigues</li>
<li>Andy Kern</li>
<li>Philipp Messer</li>
</ul>
</div>
</div>
</section>
<section id="how-to-get-help" class="slide level2">
<h2>How to get help</h2>
<ul>
<li><p>SLiM: <a href="https://groups.google.com/forum/#!forum/slim-discuss">the mailing list</a></p></li>
<li><p>msprime/tskit: <a href="https://github.com/tskit-dev/msprime/discussions">“discussions” on github</a></p></li>
<li><p>Get involved! Suggest features, write documentation, write code…</p></li>
</ul>
</section></section>
    </div>
  </div>

  <script src="https://unpkg.com/reveal.js@^4//dist/reveal.js"></script>

  // reveal.js plugins
  <script src="https://unpkg.com/reveal.js@^4//plugin/notes/notes.js"></script>
  <script src="https://unpkg.com/reveal.js@^4//plugin/search/search.js"></script>
  <script src="https://unpkg.com/reveal.js@^4//plugin/zoom/zoom.js"></script>
  <script src="https://unpkg.com/reveal.js@^4//plugin/math/math.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
        // Display the page number of the current slide
        slideNumber: true,
        // Push each slide change to the browser history
        history: true,
        // Transition style
        transition: 'none', // none/fade/slide/convex/concave/zoom
        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [
          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    </body>
</html>
